sandbox/lua/decoders/json.lua

```diff
@@ -0,0 +1,170 @@
+
+- Timestamp (string, optional, default nil)
+    string specifying json field to map to message Timestamp,
+    expects field value to be unix epoch timestamp in seconds.
+
+.. code-block:: javascript
+
+    {
+      "msg": "Start Request",
+      "event": "artemis.web.ensure-running",
+      "extra": {
+        "workspace-id": "cN907xLngi"
+      },
+      "time": "2015-05-06T20:40:05.509926234Z",
+      "severity": 1
+    }
+
+*Example Heka Configuration*
+
+.. code-block:: ini
+
+    [ArtemisLogInput]
+    type = "LogstreamerInput"
+    log_directory = "/srv/artemis/current/logs"
+    file_match = 'artemis\.log'
+    decoder = "JsonDecoder"
+
+    [JsonDecoder]
+    type = "SandboxDecoder"
+    filename = "lua_decoders/json.lua"
+
+        [JsonDecoder.config]
+        type = "artemis"
+        payload_keep = true
+        map_fields = true
+        Severity = "severity"
+
+*Example Heka Message*
+
+:Timestamp: 2015-05-06 20:40:05 -0800 PST
+:Type: artemis
+:Hostname: test.example.com
+:Pid: 0
+:UUID: 8e414f01-9d7f-4a48-a5e1-ae92e5954df5
+:Payload:
+:EnvVersion:
+:Severity: 1
+:Fields:
+    | name:"msg" value_type:STRING value_string:"Start Request"
+    | name:"event" value_type:STRING value_string:"artemis.web.ensure-running"
+    | name:"extra.workspace-id" value_type:STRING value_string:"cN907xLngi"
+    | name:"time" value_type:STRING value_string:"2015-05-06T20:40:05.509926234Z"
+--]]
+
+require "cjson"
+local util = require("util")
+
+local msg_type     = read_config("type") or "json"
+local payload_keep = read_config("payload_keep")
+local map_fields   = read_config("map_fields")
+
+local field_map = {
+    Uuid       = read_config("Uuid"),
+    Type       = read_config("Type"),
+    Logger     = read_config("Logger"),
+    Hostname   = read_config("Hostname"),
+    Severity   = read_config("Severity"),
+    EnvVersion = read_config("EnvVersion"),
+    Pid        = read_config("Pid"),
+    Timestamp  = read_config("Timestamp")
+}
+
+local field_type_map = {
+    Uuid       = "string",
+    Type       = "string",
+    Logger     = "string",
+    Hostname   = "string",
+    Severity   = "number",
+    EnvVersion = "number",
+    Pid        = "number",
+    Timestamp  = "number"
+}
+
+local msg = {
+    Type     = msg_type,
+    Payload  = nil,
+    Fields   = nil,
+    Severity = nil
+}
+
+function process_message()
+    local ok, json = pcall(cjson.decode, read_message("Payload"))
+    if not ok then
+        return -1
+    end
+
+    -- keep payload, or not
+    if payload_keep then
+        msg.Payload = read_message("Payload")
+    end
+
+    -- map fields
+    if map_fields then
+        for F, f in pairs(field_map) do
+            if type(json[f]) == field_type_map[F] then
+                if F == "Timestamp" then
+                    msg[F] = json[f] * 1e9
+                else
+                    msg[F] = json[f]
+                end
+                json[f] = nil
+            end
+        end
+    end
```

Comment:
> I think we're leaking values through here.

My bad; I guess I misunderstood what happens between decodes; I assumed the initialization before the process message function was run for every decode, but it sounds like that's not the case.

>We should add all of the possible fields in use to the msg initialization
Gotcha.

> and then be sure to explicitly nil out or set defaults for all static message fields that we don't get from the input JSON.

If I'm correctly interpreting the suggestion here, wouldn't it be easier to just set up the initial values for the msg table at the top of the process_message function at this point?